File: src/core/src/main/java/org/semachina/jena/datatype/factory/YearFactory.java
package org.semachina.jena.datatype.factory;

import com.hp.hpl.jena.vocabulary.XSD;
import org.semachina.jena.datatype.types.Year;
import org.semachina.jena.impl.SimpleRDFDatatypeImpl;

public class YearFactory extends SimpleRDFDatatypeImpl<Year> {
    public YearFactory() {
        super(XSD.gYear.getURI(), Year.class);
    }

    @Override
    public Year parseLexicalForm(String lexicalForm) throws Exception {
        return new Year(lexicalForm);
    }

    @Override
    public String toLexicalForm(Year cast) throws Exception {
        return cast.toString();
    }

    @Override
    public Object cannonicalise(Object value) {
        if (value instanceof Number) {
            return new Year(((Number) value).intValue());
        }
        return super.cannonicalise(value);
    }
}

File: src/core/src/main/scala/org/semachina/jena/datatype/factory/YearFactory.scala
package org.semachina.jena.datatype.factory

import com.hp.hpl.jena.vocabulary.XSD
import org.semachina.jena.datatype.types.Year
import org.semachina.jena.datatype.SimpleRDFDatatype

class YearFactory extends SimpleRDFDatatype[Year](
  XSD.gYear.getURI,
  { lexicalForm: String => new Year( lexicalForm ) } ) {

  override def cannonicalise(value: AnyRef): Year = {
    if (value.isInstanceOf[Number]) {
      return new Year((value.asInstanceOf[Number]).intValue)
    }
    return super.cannonicalise(value)
  }
}